module/zfs/dbuf.c

```diff
@@ -1234,11 +1224,44 @@ dbuf_read_impl(dmu_buf_impl_t *db, zio_t *zio, uint32_t flags)
 		spa_log_error(db->db_objset->os_spa, &zb);
 		zfs_panic_recover("unencrypted block in encrypted "
 		    "object set %llu", dmu_objset_id(db->db_objset));
+		DB_DNODE_EXIT(db);
+		mutex_exit(&db->db_mtx);
 		return (SET_ERROR(EIO));
 	}
 
+	/*
+	 * If we are doing a decrypting read of a block, make sure we
+	 * have decrypted the dnode associated with it so that the current
+	 * bp's MAC can be verified.
+	 */
+	if (abuf != NULL && dn->dn_objset->os_encrypted &&
+	    DMU_OT_IS_ENCRYPTED(dn->dn_bonustype) &&
+	    (flags & DB_RF_NO_DECRYPT) == 0 &&
+	    arc_is_encrypted(abuf)) {
+		SET_BOOKMARK(&zb, dmu_objset_id(db->db_objset),
+		    DMU_META_DNODE_OBJECT, 0, dn->dn_dbuf->db_blkid);
```

Comment:
This seems to be assuming that the parent dbuf is a block of dnodes, but we haven't done anything to check or verify that.  Maybe that is always the case if arc_is_encrypted(), but it seems fragile to make that assumption.  We should either make this handle different kinds of blocks, under the assumption that any encrypted buf would need to be transformed (verified), or we should check/assert that this only happens when the parent is a block of dnodes.

Reply:
`abuf` is set like this: `abuf = (dn->dn_dbuf != NULL) ? dn->dn_dbuf->db_buf : NULL;`

Am I wrong to assume this will always be the block of dnodes? Is the name confusing? This variable was originally called `dn_buf` but Brian asked me to change it.

Reply:
Ah, I misread the definition of `abuf`.  I thought that it was the buf containing the BP that we will read (i.e. `db->db_parent->db_buf`).  IMO, it would make more sense if that were the case.  Since I think that's what we care about.  If we are reading a L0 and we are pointed to by a L1 block, we should be able to assume that if the L1 block is in memory and not encrypted, then it is valid and verified.  In practice we will only call arc_untransform() here when we're reading a block that the dnode directly points to.

If you keep the current definition, it would be nice if the name contained `dnode`, to make it more clear that it's associated with the dnode, rather than (directly) with the dbuf.  e.g. `dnode_buf` or `dnode_abuf`.

Reply:
Blocks above level 0 are not encrypted. They have checksums-of-MACs which are always checked since these don't need the keys. This will only be called for level 0 dnode buffers.

Reply:
`dnode_abuf` is the most clear to me.

Reply:
OK, so maybe we should do the `dnode_abuf`, but just add a comment explaining this, and an assertion that this happens (the dnode_abuf is still encrypted) only when reading the block directly below it? something like 
```
/* When reading the topmost block of this dnode, we may not have yet verified ... 
therefore we must ....  If we are reading a lower block (that's pointed to by an 
indirect block, rather than by the dnode), we should have already verified the 
dnode's block. */
assert(db->db_level == dn->dn_nlevels-1)
```